File: Java/src/Copy List with Random Pointer.java
/**
 * Definition for singly-linked list with a random pointer.
 * class RandomListNode {
 * int label;
 * RandomListNode next, random;
 * RandomListNode(int x) { this.label = x; }
 * };
 */
public class Solution
{
  public RandomListNode copyRandomList(RandomListNode head)
  {
    if (head == null)
    {
      return null;
    }
    RandomListNode node = head;
    while (node != null)
    {
      RandomListNode nodeIns = new RandomListNode(node.label);
      RandomListNode tmp = node.next;
      node.next = nodeIns;
      nodeIns.next = tmp;
      node = tmp;
    }

    node = head;
    while (node != null)
    {
      RandomListNode nodeIns = node.next;
      if (node.random != null)
        nodeIns.random = node.random.next;
      node = node.next.next;
    }

    // detach list
    RandomListNode nodeIns = head.next;
    node = head;
    RandomListNode ret = head.next;

    while (nodeIns != null && node != null)
    {
      node.next = node.next.next;
      if (nodeIns.next == null)
      {
        break;
      }
      nodeIns.next = nodeIns.next.next;

      nodeIns = nodeIns.next;
      node = node.next;
    }
    return ret;

  }
}

File: Python/Copy List with Random Pointer.py
# Definition for singly-linked list with a random pointer.
# class RandomListNode:
#     def __init__(self, x):
#         self.label = x
#         self.next = None
#         self.random = None

class Solution:
    # @param head, a RandomListNode
    # @return a RandomListNode
    def copyRandomList(self, head):
        if not head: return None
        node = head
        while node:
            nodeIns = RandomListNode(node.label)
            temp = node.next
            node.next = nodeIns
            nodeIns.next = temp
            node = temp
        
        node = head
        while node:
            nodeIns = node.next
            if node.random:
                nodeIns.random = node.random.next
            node = node.next.next
        
        node = head
        nodeIns = head.next
        ret = head.next
        while node and nodeIns:
            node.next = node.next.next
            if not nodeIns.next:
                break
            nodeIns.next = nodeIns.next.next
            nodeIns = nodeIns.next
            node = node.next
        return ret
            